File: Python/Chapter4/FindMaximumSubArray.py

def FindMaximumSubArray(A, start, end):
    if end - start > 1:
        mid = (start + end)/2
        left_low_index, left_high_index, left_sum = FindMaximumSubArray(A, start, mid)
        right_low_index, right_high_index, right_sum = FindMaximumSubArray(A, mid, end)
        cross_low_index, cross_high_index, cross_sum = FindMaxCrossingSubArray(A, start, mid, end)
        if left_sum >= right_sum and left_sum >= cross_sum:
            return (left_low_index, left_high_index, left_sum)
        elif right_sum >= left_sum and right_sum >= cross_sum:
            return (right_low_index, right_high_index, right_sum)
        else:
            return (cross_low_index, cross_high_index, cross_sum)
    else:
        return (start, end, A[start])

def FindMaxCrossingSubArray(A, start, mid, end):
    sum = 0
    max_low_sum = A[mid - 1]
    max_low_index = mid - 1
    for i in range(mid - 1, start - 1):
        sum += A[i]
        if sum > max_low_sum:
            max_low_sum = sum
            max_low_index = i
    sum = 0
    max_high_sum = A[mid]
    max_high_index = mid
    for i in range(mid, end):
        sum += A[i]
        if sum > max_high_sum:
            max_high_sum = sum
            max_high_index = i
    return (max_low_index, max_high_index, max_low_sum + max_high_sum)

A = [13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7]
print("low_index:%d, high_index:%d, sum:%d" % FindMaximumSubArray(A, 0, len(A)))

File: C++/Chapter4/FindMaximumSubArray.cpp
#include "FindMaximumSubArray.h"
#include <stdio.h>

using namespace std;

std::tuple<int, int, int> FindMaximumSubArray(int* a, int start, int end)
{
	if (end - start > 1)
	{
		int mid = (start + end) / 2;
		auto left = FindMaximumSubArray(a, start, mid);
		auto right = FindMaximumSubArray(a, mid, end);
		auto cross = FindMaxCrossingSubArray(a, start, mid, end);
		if (get<2>(left) >= get<2>(right) && get<2>(left) >= get<2>(cross))
		{
			return left;
		}
		else if (get<2>(right) >= get<2>(left) && get<2>(right) >= get<2>(cross))
		{
			return right;
		}
		else
		{
			return cross;
		}
	}
	else
	{
		return tuple<int, int, int>(start, end, a[start]);
	}
}

std::tuple<int, int, int> FindMaxCrossingSubArray(int* a, int start, int mid, int end)
{
	int max_low_sum = a[mid - 1];
	int max_low_index = mid - 1;
	int sum = 0;
	for (int i = mid - 1; i >= 0; i--)
	{
		sum += a[i];
		if (sum > max_low_sum)
		{
			max_low_sum = sum;
			max_low_index = i;
		}
	}
	int max_high_sum = a[mid];
	int max_high_index = mid;
	sum = 0;
	for (int i = mid; i < end; i++)
	{
		sum += a[i];
		if (sum > max_high_sum)
		{
			max_high_sum = sum;
			max_high_index = i;
		}
	}
	return tuple<int, int, int>(max_low_index, max_high_index, max_low_sum + max_high_sum);
}


void TestFindMaximumSubArray()
{
	int a[] = { 13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7 };
	int len = sizeof(a) / sizeof(int);
	auto result = FindMaximumSubArray(a, 0, len);
	printf("left_index:%d, right_index:%d, max_sum:%d\n", get<0>(result), get<1>(result), get<2>(result));
}
